Egress SDS Perf Test

nschhina opened the pull request:
closes https://github.com/istio/istio/issues/25653

istio-testing commented:
@nschhina: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_tools | 229a904b9cb3b969a118b9d863279d124dee02a3 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_tools/1108/lint_tools/1572) | `/test lint_tools`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nschhina: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_tools | 9d9faf01ee9e9d069286b3ea31f62f953b788cf5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_tools/1108/lint_tools/1579) | `/test lint_tools`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nschhina: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_tools | 1ab8f5ab7f598a04acc44bd1e3f5fb2f24995c5c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_tools/1108/lint_tools/1588) | `/test lint_tools`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

JimmyCYJ commented on perf/security/sds-tests/egress-sds/README.md:
````diff
@@ -0,0 +1,46 @@
+To download a specific version of istioctl and deploy the test worloads using that istioctl binary.
+
+    ```bash
+    NUM=100 CLUSTER=gke_istio-security-testing_us-central1-a_istio-testing ./setup_test.sh 1.5.1 pre-release
````
Change 1.5.1 to 1.7.0? We don't support egress SDS on 1.5.1

JimmyCYJ commented on perf/security/sds-tests/egress-sds/request_script.sh:
```diff
@@ -0,0 +1,31 @@
+  fi
+  num_curl=$((num_curl+1))
+  echo "$(date +"%Y-%m-%d %H:%M:%S:%3N") Out of ${num_curl} curl, ${num_succeed} succeeded."
+  sleep 1
```
Can we make it 0.1, so the qps is 10/sec. 1/sec is low for a perf test

JimmyCYJ commented on perf/security/sds-tests/egress-sds/setup_destinationrule.sh:
```diff
@@ -0,0 +1,46 @@
+    - port:
+        number: 443
+      tls:
+        mode: MUTUAL
```
README says this is TLS gateway, I think we are using mTLS. Could you update README?

JimmyCYJ commented on perf/security/sds-tests/egress-sds/setup_destinationrule.sh:
```diff
@@ -0,0 +1,46 @@
+        sni: "${host}"
+EOF
+
+#    istioctl experimental wait -n istio-system destinationrule "${dr_name}" --timeout=60s
```
uncomment this command?

nschhina replied:
I tried using experimental wait but since it's not production ready I was seeing some longer timeouts so I decided to comment this out, I'll remove this completely actually.

JimmyCYJ commented on perf/security/sds-tests/egress-sds/setup_gateway.sh:
```diff
@@ -0,0 +1,68 @@
+          mode: ISTIO_MUTUAL
+          sni: "${host_nginx}"
+EOF
+#    istioctl experimental wait -n "${ns}" gateway "${gateway_name}" --timeout=60s
```
uncomment these two lines?

nschhina replied:
same as above.

carolynhu commented on perf/security/sds-tests/egress-sds/README.md:
```diff
@@ -0,0 +1,46 @@
+The test sets up a number of TLS egress gateways for a group of nginx services.
+The test creates a group of sleep pods, where each sleep sends HTTP requests to a nginx
+service periodically. For example, sleep-1 sends HTTP requests to my-nginx-1,
+and sleep-2 sends HTTPS requests to my-nginx-2.example.com. The sleep pods are deployed in clientns and nginx are deployed
```
nit: make the name `my-nginx-2.example.com`  to be consistent with `my-nginx-1`

carolynhu commented on perf/security/sds-tests/egress-sds/README.md:
```diff
@@ -0,0 +1,46 @@
+
+This is a TLS egress gateway SDS test.
+The test sets up a number of TLS egress gateways for a group of nginx services.
+The test creates a group of sleep pods, where each sleep sends HTTP requests to a nginx
```
nit: sleeping

nschhina replied:
these are sleep pods not sleeping pods, sleep is more accurate to describe tem

nschhina commented:
> @nschhina Sorry, It's better for you to first check your README to make it has less wording errors and make it consistent with what's in your code. It would be much more easier for reviewing. Thank you very much!

What part is not clear/doesn't match the actual test? I've modified the README that already existed for ingress sds tests and added nginx components. This test is just the same test but instead of ingress it uses egress gateway deployment

istio-testing commented:
@nschhina: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_tools | c982752b0a6aa23d2b130d6beb48bb65cf930ebc | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_tools/1108/lint_tools/1612) | `/test lint_tools`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

carolynhu commented on perf/security/sds-tests/egress-sds/README.md:
```diff
@@ -0,0 +1,50 @@
+The test sets up a number of TLS egress gateways for a group of nginx services.
+The test creates a group of sleep pods, where each sleep sends HTTP requests to an nginx
+service periodically. For example, sleep-1 sends HTTP requests to my-nginx-1,
+and sleep-2 sends HTTPS requests to my-nginx-2. The sleep pods are deployed in clientns namespace and nginx services are deployed
```
if `clientns` is the name of your namespace, please add quotes to it to differentiate from the normal text.

carolynhu commented on perf/security/sds-tests/egress-sds/README.md:
```diff
@@ -0,0 +1,50 @@
+
+## To run the SDS test originating mTLS at gateway
+
+- Create a GKE cluster and set it as the current cluster.
+Here this test is run on the cluster *istio-testing*
+on GCP project *istio-security-testing*.
+You may use `kubectl config current-context` to confirm that your newly created cluster
+is set as the current cluster.
```
These description may not be necessary. You can just saying that `Create a GEK cluster and run the test against your created cluster`.

carolynhu commented on perf/security/sds-tests/egress-sds/README.md:
````diff
@@ -0,0 +1,50 @@
+    NUM=100 CLUSTER=gke_istio-security-testing_us-central1-a_istio-testing ./setup_test.sh
+    ```
+
+The logs for sleep containers shows number of requests that are sent to nginx and number of successful responses received.
````
nit: show

carolynhu commented on perf/security/sds-tests/egress-sds/README.md:
````diff
@@ -0,0 +1,50 @@
+    ```
+
+The logs for sleep containers shows number of requests that are sent to nginx and number of successful responses received.
+The logs for istio-egressgateway also show the GET requests that are routed from sleep pod's in clientns namespace to corresponding
````
quote `clientns`

carolynhu commented on perf/security/sds-tests/egress-sds/cleanup.sh:
```diff
@@ -0,0 +1,35 @@
+  do
+    kubectl -n istio-system delete "${s}"
+  done
+    for s in $(kubectl -n istio-system get secrets -oname | grep "nginx-server-certs-*")
```
fix indentation here?

nschhina replied:
Indentation looks good to me?

JimmyCYJ replied:
I think `for s in $(kubectl -n istio-system get secrets -oname | grep "nginx-server-certs-*")` could start at a new line (#30) and remove the indentation at the beginning?

carolynhu commented on perf/security/sds-tests/egress-sds/setup_nginx.sh:
```diff
@@ -0,0 +1,110 @@
+    local id="${1:?"please specify the gateway id"}"
+    local ns="${2:?"please specify the namespace"}"
+    local cs="${3:?"please specify the cluster"}"
+    local wd="${4:?"please specify the cert path"}"
```
`wd` here is for?

nschhina replied:
wd is for working directory and cert path is in working directory. I will change this up though

carolynhu commented on perf/security/sds-tests/egress-sds/setup_nginx.sh:
```diff
@@ -0,0 +1,110 @@
+function setup_nginx () {
+    local id="${1:?"please specify the gateway id"}"
+    local ns="${2:?"please specify the namespace"}"
+    local cs="${3:?"please specify the cluster"}"
```
`ns` can be the short of namespace, well, `cluster` is short enough, just using `cluster` here, instead of using hard to interprete `cs`?

nschhina replied:
Honestly this is not by choice. I kept it as `cs` just to be consistent with what we have in ingress-sds perf tests

JimmyCYJ replied:
Yes, we use cs in other security tests as well. How about we keep it consistent here? We can rename variables for all the security tests in a followup PR.

carolynhu replied:
yes, sure!

carolynhu commented on perf/security/sds-tests/egress-sds/README.md:
```diff
@@ -0,0 +1,50 @@
+The test creates a group of sleep pods, where each sleep sends HTTP requests to an nginx
+service periodically. For example, sleep-1 sends HTTP requests to my-nginx-1,
+and sleep-2 sends HTTPS requests to my-nginx-2. The sleep pods are deployed in clientns namespace and nginx services are deployed
+in a namespace called mesh-external".
```
fix mesh-external"?
